tensorflow/cc/framework/gradients.cc

```diff
@@ -165,6 +165,27 @@ Status SymbolicGradientBuilder::Initialize() {
   // handle computing gradients over small subgraphs from a very large graph.
   pending_.resize(scope_.graph()->num_node_ids(), 0);
   {
+    // If a node is not reachable when doing a BFS from the outputs
+    // we don't expect it to receive a backpropagated gradient.
+    // It will not be counted in num_expected_backprops.
+    std::vector<bool> reachable_nodes(scope_.graph()->num_node_ids(), false);
```

Comment:
This function is already quite long, can you break this out into a new function? Something like GetReachableNodes().